xarray/core/coordinates.py

```diff
@@ -796,6 +789,113 @@ def _ipython_key_completions_(self):
+
+    @property
+    def dims(self) -> Frozen[Hashable, int]:
+        # TODO is there a potential bug here? What happens if a dim is only present on data variables?
+        return self._data.dims
+
+    @property
+    def dtypes(self) -> Frozen[Hashable, np.dtype]:
+        """Mapping from coordinate names to dtypes.
+
+        Cannot be modified directly, but is updated when adding new variables.
+
+        See Also
+        --------
+        Dataset.dtypes
+        """
+        return Frozen({n: v.dtype for n, v in self._data._coord_variables.items()})
+
+    @property
+    def variables(self) -> Mapping[Hashable, Variable]:
+        return Frozen(self._data._coord_variables)
+
+    def __getitem__(self, key: Hashable) -> DataArray:
+        if key not in self._data._coord_variables:
+            raise KeyError(key)
+        item = self._data[key]  # type: ignore[index]  # see https://github.com/pydata/xarray/issues/8836
+
+        # TODO perhaps instead use an internal `DataTree` getter method which always returns DataArrays here?
+        return cast("DataArray", item)
+
+    def to_dataset(self) -> Dataset:
+        """Convert these coordinates into a new Dataset"""
+        return self._data.to_dataset()._copy_listed(self._names)
+
+    def _update_coords(
+        self, coords: dict[Hashable, Variable], indexes: Mapping[Any, Index]
+    ) -> None:
+
+        # create updated node (`.to_dataset` makes a copy so this doesn't modify in-place)
+        node_ds = self._data.to_dataset(inherited=False)
+        node_ds.coords._update_coords(coords, indexes)
+
+        from xarray.core.datatree import _check_alignment
+
+        # check consistency *before* modifying anything in-place
+        # TODO can we clean up the signature of _check_alignment to make this less awkward?
+        if self._data.parent is not None:
+            parent_ds = self._data.parent._to_dataset_view(
+                inherited=True, rebuild_dims=False
+            )
+        else:
+            parent_ds = None
+        _check_alignment(self._data.path, node_ds, parent_ds, self._data.children)
+
+        # assign updated attributes
+        coord_variables = {
+            k: v for k, v in node_ds.variables.items() if k in node_ds._coord_names
+        }
+        self._data._node_coord_variables = coord_variables
+        self._data._node_dims = node_ds._dims
+        self._data._node_indexes = node_ds._indexes
+
+    def _drop_coords(self, coord_names):
+        # should drop indexed coordinates only
+        for name in coord_names:
+            del self._data._node_coord_variables[name]
+            del self._data._node_indexes[name]
+
+    def __delitem__(self, key: Hashable) -> None:
+        if key in self:
+            del self._data[key]  # type: ignore[arg-type]  # see https://github.com/pydata/xarray/issues/8836
+        else:
+            raise KeyError(
+                f"{key!r} is not in coordinate variables {tuple(self.keys())}"
+            )
```

Comment:
KeyError has special formatting, so you can only pass the `key` here:
```suggestion
            raise KeyError(key)
```